### PR TITLE
Stream Impersonation

### PR DESCRIPTION
**Pull request by gokulavasan**
Ready for review. Tested different scenarios (except bulk upload) on a secure cluster with impersonation. 

JIRA : https://issues.cask.co/browse/CDAP-6435
Build : http://builds.cask.co/browse/CDAP-DUT4465
ITN : http://builds.cask.co/browse/CDAP-ITM-25

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java [hunk 1]
```diff
@@ -63,25 +70,48 @@ public void cleanAll() throws Exception {
     List<NamespaceMeta> namespaces = namespaceQueryAdmin.list();
 
     for (NamespaceMeta namespace : namespaces) {
-      Location streamBaseLocation =
+      final Location streamBaseLocation =
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: this should go inside callable right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java [hunk 2]
```diff
@@ -63,25 +70,48 @@ public void cleanAll() throws Exception {
...
         namespacedLocationFactory.get(Id.Namespace.from(namespace.getName())).append(streamBaseDirPath);
-      if (!streamBaseLocation.exists()) {
+      UserGroupInformation ugi = impersonator.getUGI(new NamespaceId(namespace.getName()));
+      boolean exists = ImpersonationUtils.doAs(ugi, new Callable<Boolean>() {
```
Comment (gsps1) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java:80]: `impersonator.doAs`

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java (hunk 1 above)
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: why? it is not doing any filesystem operation.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java (hunk 2 above)
Comment (gokulavasan) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java:80]: I want to get the ugi and use it for multiple operations. Hence I am using the doAs method in ImpersonationUtils.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/security/ImpersonationUtils.java [hunk 3]
```diff
@@ -40,6 +40,9 @@ private ImpersonationUtils() { }
    */
   public static <T> T doAs(UserGroupInformation ugi, final Callable<T> callable) throws Exception {
     try {
+      if (ugi == null) {
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/security/ImpersonationUtils.java]: this shouldn't be needed. UGI shouldn't be passed in as null

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/security/Impersonator.java [hunk 4]
```diff
@@ -57,8 +58,14 @@ public Impersonator(CConfiguration cConf, UGIProvider ugiProvider,
     if (!kerberosEnabled) {
       return callable.call();
     }
+    return ImpersonationUtils.doAs(getUGI(namespaceId), callable);
+  }
+
+  public UserGroupInformation getUGI(NamespaceId namespaceId) throws IOException {
+    if (!kerberosEnabled) {
+      return null;
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/security/Impersonator.java]: return UserGroupInformation.getCurrentUser()

**Comment by gokulavasan** [comment]
Build is green 🍏

**Comment by anwar6953** [comment]
Changes lgtm.
Just wondering about the impact on stream notifications, and any necessary changes. Perhaps can be done in a separate PR later, but we should figure out if that's necessary at all.
This PR 👍

**Comment by gokulavasan** [comment]
@anwar6953 Discussed offline. Looks like we don't need any changes to StreamNotifications. But ITN tests will let us know soon.

**Comment by anwar6953** [comment]
StreamNotifications service doesn't interact directly with stream files; it is based off of kafka notifications and stream metrics, so it doesn't need impersonation.

**Pull request merged**